osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -89,28 +103,108 @@ public void updateFirmware(final CorrelationIds ids,
         this.findOrganisation(ids.getOrganisationIdentification());
         final Device device = this.findActiveDevice(ids.getDeviceIdentification());
 
-        this.osgpCoreRequestMessageSender.send(
+        if (device instanceof Ssld) {
+            this.createSsldPendingFirmwareUpdateRecord(ids, firmwareUpdateMessageDataContainer.getFirmwareModuleData(),
+                    firmwareUpdateMessageDataContainer.getFirmwareUrl());
+        }
+
+        this.osgpCoreRequestMessageSender.sendWithScheduledTime(
                 new RequestMessage(ids,
                         this.domainCoreMapper.map(firmwareUpdateMessageDataContainer,
                                 org.opensmartgridplatform.dto.valueobjects.FirmwareUpdateMessageDataContainer.class)),
                 messageType, messagePriority, device.getIpAddress(), scheduleTime);
     }
 
+    private void createSsldPendingFirmwareUpdateRecord(final CorrelationIds ids,
+            final FirmwareModuleData firmwareModuleData, final String firmwareUrl) {
+        try {
+            final String firmwareFilename = getFirmwareFilename(firmwareUrl);
+
+            final List<FirmwareFile> firmwareFiles = this.firmwareFileRepository.findByFilename(firmwareFilename);
+            Assert.isTrue(firmwareFiles.size() == 1, "Expected 1 firmware file for filename: " + firmwareFilename);
+            final FirmwareFile firmwareFile = firmwareFiles.get(0);
+
+            final Map<FirmwareModule, String> firmwareModuleVersions = firmwareFile.getModuleVersions();
+            Assert.isTrue(firmwareModuleVersions.size() == 1,
+                    "Expected 1 firmware module for: " + firmwareModuleVersions);
+            final String firmwareVersion = getFirmwareVersion(firmwareModuleVersions);
+            final FirmwareModuleType firmwareModuleType = getFirmwareModuleType(firmwareModuleVersions);
+
+            SsldPendingFirmwareUpdate ssldPendingFirmwareUpdate = new SsldPendingFirmwareUpdate(
+                    ids.getDeviceIdentification(), firmwareModuleType, firmwareVersion,
+                    ids.getOrganisationIdentification(), ids.getCorrelationUid());
+            ssldPendingFirmwareUpdate = this.ssldPendingFirmwareUpdateRepository.save(ssldPendingFirmwareUpdate);
+
+            LOGGER.info("Saved pending fimware update record for SSLD: {}, {}", ids.getDeviceIdentification(),
+                    ssldPendingFirmwareUpdate);
+        } catch (final Exception e) {
+            LOGGER.error("Caugth exception when creating pending firmware update record for SSLD: {}",
+                    ids.getDeviceIdentification(), e);
+        }
+    }
+
+    private static String getFirmwareFilename(final String firmwareUrl) {
+        final String[] split = firmwareUrl.split("/");
+        Assert.isTrue(split.length >= 1, "Splitting URL on / failed!");
+        final String firmwareFilename = split[split.length - 1];
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "firmwareFilename". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1488)